cmd/av/commit_split.go

```diff
@@ -0,0 +1,142 @@
+		repo, err := getRepo()
+		if err != nil {
+			return err
+		}
+		if clean, err := repo.CheckCleanWorkdir(); err != nil {
+			return err
+		} else if !clean {
+			return errors.New("the workdir is not clean")
+		}
+
+		// Ignore errors to support a detached HEAD.
+		currentBranchName, _ := repo.CurrentBranchName()
+		currentCommitOID, err := repo.RevParse(&git.RevParse{Rev: "HEAD"})
+		if err != nil {
+			return errors.Errorf("cannot get the current commit object: %v", err)
+		}
+
+		// From here, we use detached HEAD, so that even if something goes wrong or user
+		// aborts the operation in the middle, the original branch is intact.
+		if err := splitCommit(repo, currentBranchName, currentCommitOID); err != nil {
+			commitSplitAbortMessage(currentBranchName, currentCommitOID)
+			return err
+		}
+		return nil
+	},
+}
+
+func splitCommit(repo *git.Repo, currentBranchName, currentCommitOID string) error {
+	if _, err := repo.Git("switch", "--detach", currentCommitOID); err != nil {
+		return err
+	}
+	if _, err := repo.Git("reset", "--mixed", "HEAD~"); err != nil {
+		return err
+	}
+	if _, err := repo.Git("add", "--intent-to-add", repo.Dir()); err != nil {
+		return err
+	}
+
+	for {
+		if clean, err := repo.CheckCleanWorkdir(); err != nil {
+			return err
+		} else if clean {
+			break
+		}
+
+		if _, err := repo.Run(&git.RunOpts{
+			Args:        []string{"add", "--patch"},
+			ExitError:   true,
+			Interactive: true,
+		}); err != nil {
+			return err
+		}
+
+		if hasStagedChange, err := repo.HasChangesToBeCommitted(); err != nil {
+			return err
+		} else if !hasStagedChange {
+			return errors.New("nothing is selected to commit")
+		}
+
+		if _, err := repo.Run(&git.RunOpts{
+			// Add --verbose to show the diffs to be committed.
+			Args:        []string{"commit", "--verbose", "--reedit-message", currentCommitOID},
+			ExitError:   true,
+			Interactive: true,
+		}); err != nil {
+			return err
+		}
+	}
+
+	if currentBranchName != "" {
+		newCommitOID, err := repo.RevParse(&git.RevParse{Rev: "HEAD"})
+		if err != nil {
+			return errors.Errorf("cannot get the resulting commit object: %v", err)
+		}
+		if err := repo.UpdateRef(&git.UpdateRef{
+			Ref: "refs/heads/" + currentBranchName,
+			Old: currentCommitOID,
+			New: newCommitOID,
+			// Add this change to reflog so that the user can revert back to the
+			// original state.
+			CreateReflog: true,
+		}); err != nil {
+			return errors.Errorf("cannot update the branch %s to the new commit %s: %v", currentBranchName, newCommitOID, err)
+		}
+		// At this point, the HEAD is still a detached HEAD. Check out the branch.
+		// repo.CheckoutBranch errors out if the repository is at the detached head.
+		// We have to run git checkout in other ways.
+		if _, err := repo.Git("switch", currentBranchName); err != nil {
+			return errors.Errorf("cannot switch to the original branch: %v", err)
+		}
+
+		// TODO: We should rebase the stacks after split.
```

Comment:
can you create an issue for this in GH?

Reply:
At the very least, can you print out a message to the user that says
```
Run `av stack sync` to sync your stack if necessary.
```

Reply:
Done.